prospero-common/src/main/java/org/wildfly/prospero/actions/ApplyStageBackup.java

```diff
@@ -0,0 +1,216 @@
+            ProsperoLogger.ROOT_LOGGER.debug("Creating backup record in " + backupRoot);
+        }
+
+        if (!Files.exists(backupRoot)) {
+            if (ProsperoLogger.ROOT_LOGGER.isTraceEnabled()) {
+                ProsperoLogger.ROOT_LOGGER.trace("Creating backup directory in " + backupRoot);
+            }
+
+            Files.createDirectories(backupRoot);
+        } else if (!Files.isDirectory(backupRoot) || !Files.isWritable(backupRoot)) {
+            throw new RuntimeException(String.format(
+                    "Unable to create backup in %s. It is not a directory or is not writable.",
+                    backupRoot
+            ));
+        }
+
+        final File[] files = backupRoot.toFile().listFiles();
+        if (files != null) {
+            for (File file : files) {
+                if (ProsperoLogger.ROOT_LOGGER.isTraceEnabled()) {
+                    ProsperoLogger.ROOT_LOGGER.trace("Removing existing backup files: " + file);
+                }
+
+                FileUtils.forceDelete(file);
+            }
+        }
+    }
+
+    /**
+     * add all the files in the server to cache
+     *
+     * @throws IOException - if unable to backup the files
+     */
+    public void recordAll() {
+        try {
+            Files.walkFileTree(serverRoot, new SimpleFileVisitor<>() {
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    if (dir.equals(backupRoot)) {
+                        return FileVisitResult.SKIP_SUBTREE;
+                    } else {
+                        final Path relative = serverRoot.relativize(dir);
+                        Files.createDirectories(backupRoot.resolve(relative));
+                        return FileVisitResult.CONTINUE;
+                    }
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    final Path relative = serverRoot.relativize(file);
+                    if (relative.startsWith(Path.of(".installation", ".git"))) {
+                        // when using hardlinks, we need to remove the file and copy the new one in it's place otherwise both would be changed.
+                        // the git folder is manipulated by jgit and we can't control how it operates on the files
+                        // therefore we need to copy the files upfront rather than hardlinking them
+                        Files.copy(file, backupRoot.resolve(relative));
+                    } else {
+                        // we try to use hardlinks instead of copy to save disk space
+                        // fallback on copy if Filesystem doesn't support hardlinks
+                        try {
+                            Files.createLink(backupRoot.resolve(relative), file);
+                        } catch (UnsupportedEncodingException e) {
```

Comment:
Is that the exception that's gonna be thrown when unable to create a hardlink? I would expect UnsupportedOperationException.

Reply:
We may try to check explicitly if the paths are on the same filesystem and revert to copy if they are not, to avoid the exception. It would not simplify the code, just avoided the exception in one predictable case. No big deal though.

Reply:
good catch, thanks. That should be `UnsupportedOperationException`. 

> We may try to check explicitly if the paths are on the same filesystem.

That would assume the installation of WildFly itself is split between different filesystems, Is that something we support/anticipate?

Reply:
Hmm, probably not :+1: